text/0000-router-service.md

```diff
@@ -0,0 +1,252 @@
+    this.replaceWith(transition.to.name, fallbackPersonId);
+  });
+}
+
+// If personId 5 is invalid, and the user visits /person/5/detail, they will get
+// redirected to /person/0/detail. And /person/5 will get redirected to /person/0.
+```
+
+
+### Actively discourage use of private API
+
+This RFC provides public API for doing the things people have become accustomed to doing via private properties on `transition`. To eliminate confusion over the correct way, we should hide all the private API away behind symbols, and provide deprecation warnings per our usual release policy around breaking "widely-used private APIs"
+
+
+# Drawbacks
+
+This RFC suggests only two small deprecations that are unlikely to effect many apps, so the API-churn burden may appear low. However, we know that use of the private APIs we're deliberately disabling is widespread, so users will experience churn. We can provide our usual deprecation cycle to give them early warning, but it still imposes some cost.
+
+This RFC doesn't attempt to change the existing and fairly rich semantics for initiating transitions. For example, you can pass either models or IDs, and those have subtle semantic differences. I think an ideal rewrite would also change the semantics of the route hooks and transitionTo to simplify that area.
+
+# Alternatives
+
+## Less Churn
+
+We could adopt some of the existing broadly used APIs as de-facto public. This avoids churn, but imposes a complexity burden on every new learner, who needs to be told "this is a weird API, but it's what we're stuck with".
+
+## Semver Lawyering
+
+I'm interepreting router.js's public/private documentation as out-of-scope for Ember's semver. The fact that we pass an instance of router.js's Transition as our `transition` argument is not documented. An alternative interpretation is that we need to continue supporting those methods marked as public in router.js's docs.
+
+## Optional Helpers
+
+I didn't propose shipping `is-active` and `url-for` template helpers -- I merely showed that they're easy to build using the router service. But we should arguably just ship them as part of the framework too.
+
+## Branching Route Hierarchies
+
+I am implicitly assuming we will only ever have linear route hierarchies, where a given route has at most one child. I can imagine eventually wanting a way to support branching route hierarchies, where each branch can transition independently. I'm not trying to account for that future.
+
+## Route.parentRoute
+
+This RFC makes it possible for a route to determine its parent's name dynamically via public API, and thus access its parent's model/params/controller:
+
+```js
+beforeModel(transition) {
+  const parentInfo = transition.to.find(info => info.name === this.routeName).parent;
+  const parentModel = this.modelFor(parentInfo.name);
+}
+```
+
+However, this pattern feels awkward, and I think it justifies just adding a public `parentRouteName()` method to `Route` that would simplify to:
```

Comment:
It would be useful to have every route in the hierarchy available via a public API. We use handlerInfos to get this information for tracking purposes right now.